### PR TITLE
8269404: Base64 Encoding optimization enhancements for x86 using AVX-512

### PR DESCRIPTION
**Pull request by asgibbons**
Enhance the Base64 Encode intrinsic for x86 using AVX-512 to get better performance. Also allow for performance improvement on non-AVX-512 enabled platforms.

Added AVX-512 code for encoding Base64 blocks, including slight improvements for non-AVX-512 x86 platforms.

Running the Base64Decode benchmark, this change increases decode performance by an average of 2.6x with a maximum 9.7x for buffers > ~20k.  The numbers are given in the table below.

**Base Score** is without intrinsic support, **Optimized Score** is using this intrinsic, and **Gain** is **Base** / **Optimized**.

Benchmark Name | Base Score | Optimized Score | Gain
-- | -- | -- | --
testBase64Encode size 1 | 8.10 | 8.04 | 1.01
testBase64Encode size 2 | 8.51 | 8.43 | 1.01
testBase64Encode size 3 | 11.08 | 10.72 | 1.03
testBase64Encode size 6 | 13.98 | 13.12 | 1.07
testBase64Encode size 7 | 14.44 | 13.38 | 1.08
testBase64Encode size 9 | 15.44 | 14.37 | 1.07
testBase64Encode size 10 | 16.13 | 14.97 | 1.08
testBase64Encode size 48 | 27.14 | 23.23 | 1.17
testBase64Encode size 512 | 123.86 | 30.75 | 4.03
testBase64Encode size 1000 | 224.42 | 37.71 | 5.95
testBase64Encode size 20000 | 4202.11 | 430.16 | 9.77

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8269404](https://bugs.openjdk.java.net/browse/JDK-8269404): Base64 Encoding optimization enhancements for x86 using AVX-512

### Reviewers
 * [Vladimir Kozlov](https://openjdk.java.net/census#kvn) (@vnkozlov - **Reviewer**) ⚠️ Review applies to c4cbe97d9f4a95b0508e8a13bd676dc1f2c4b535
 * [Sandhya Viswanathan](https://openjdk.java.net/census#sviswanathan) (@sviswa7 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4601/head:pull/4601` \
`$ git checkout pull/4601`

Update a local copy of the PR: \
`$ git checkout pull/4601` \
`$ git pull https://git.openjdk.java.net/jdk pull/4601/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4601`

View PR using the GUI difftool: \
`$ git pr show -t 4601`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4601.diff">https://git.openjdk.java.net/jdk/pull/4601.diff</a>

</details>

### CODE REVIEW

**Comment by vnkozlov** [comment]
@asgibbons if you are still working on this, please convert it to `draft` PR. There is `Convert to draft` link under `Reviewers`:
```
Reviewers
No reviews
Still in progress? Convert to draft
```
Otherwise it seems like it is ready for review and we got RFR emails.

**Comment by asgibbons** [comment]
Thanks, @vnkozlov.  I have tested this on all the platforms I have available to me, and it has been internally reviewed.  I'm a novice when it comes to this process and I have read the contributing guidelines, but I'm at a loss as to how draft PRs are used.

I believe this is the change that will ultimately be integrated, so in fact I am looking for reviews.  Can you enlighten me please?

**Review by vnkozlov — COMMENTED** [review]
At first I thought you would add more changes in Java lib and Hotspot runtime.
But you simply optimized assembler code for existing Encode intrinsic.

This looks good to me but I need to test it.

**Comment by vnkozlov** [comment]
compiler/intrinsics/base64/TestBase64.java failed with wrong result on Windows (avx512):
```
----------System.out:(1/17)----------
5000 iterations
----------System.err:(17/1502)----------
java.lang.RuntimeException: Assertion failed! The result is not same as expected
 result: ICAgICAgICAxNiBRICAgICAgICAgICAgsTsA6igAoigAoigAoigAoj0QIHk= expected: ICAgICAgICAxNiBRICAgICAgICAgICAgMzMgaCAgICAgICAgICAgIDUwIHk=
	at compiler.intrinsics.base64.TestBase64.assertEqual(TestBase64.java:238)
	at compiler.intrinsics.base64.TestBase64.test0(TestBase64.java:151)
	at compiler.intrinsics.base64.TestBase64.main(TestBase64.java:72)
```

**Comment by asgibbons** [comment]
@vnkozlov Looks like I had a shift count wrong for URL encoding.  Tested it and it should now work.  Sorry for the churn.

**Review by vnkozlov — APPROVED** [review]
Testing passed.

You need second review.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp
Comment (sviswa7) [inline src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:5672]: vzeroupper is not happening in all cases. There are exits to  L_process3.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp
Comment (asgibbons) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:5463]: There are VL/BW instructions in the loop.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp
Comment (asgibbons) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:5480]: There is no form of vpbroadcastd that takes a Register as a second argument.  Only evpbroadcastd.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp
Comment (asgibbons) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:5487]: Same as above.  No form of vpbroadcastd takes a Register as second argument.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp
Comment (asgibbons) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:5432]: I'll keep this in mind for the future.  Thanks.

**Review by sviswa7 — APPROVED** [review]

**Comment by sviswa7** [comment]
The patch looks good to me.

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp
Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:5463]: Only ones are evpbroadcastd

File: src/hotspot/cpu/x86/stubGenerator_x86_64.cpp
Comment (jatin-bhateja) [reply src/hotspot/cpu/x86/stubGenerator_x86_64.cpp:5480]: FTR we do have 256 bit vex encoded instruction to  broadcast 32bit value into ymm. 

VEX.256.66.0F38.W0 18 /r VBROADCASTSS ymm1,xmm2

https://www.felixcloutier.com/x86/vbroadcast

It seems to operates in floating point domain so sequence may incur domain change penalty, but still we can save dependency on AVX512 feature in a purely AVX2 code blob.

**Pull request closed without merging**
